### ISSUE TITLE
Option to maintain source file structure and symbol names

### ISSUE THREAD

**Issue by ribbanya** [InusualZ/dadosod#3]
Hi, in [melee](https://github.com/doldecomp/melee) we are looking for a way to dump the binary we've built back to asm. However, with the current implementation of your disassembly, we lose our file splits and symbol names in the process. We would like to preserve the file tree of our source files in the output asm, and use our symbol names (which are present in the `.elf`) instead of symbols generated based on address.

I see that "Library Splitting" is a planned feature in your README. What would be involved in implementing these features? I wouldn't mind writing them myself, but I'm not familiar with Rust or with your codebase so it would take me some time to adjust.

Thanks for your consideration.

**Comment by InusualZ** [comment]
You are asking for a few things
 - Symbol Names - I think this can be implemented easily by giving the program a map file and he take the symbol name from there.
 - File Tree for the produced ASM - This one would be a little bit tricky. How do I get the folder structure? How do I split symbols by file?
 
Also this doesn't produce a shiftable  ASM (Although that can be dealt by giving me a full MAP)

**Comment by ribbanya** [comment]
> * File Tree for the produced ASM - This one would be a little bit tricky. How do I get the folder structure? How do I split symbols by file?

It's pretty roundabout and annoying to get. The map file contains the names of the object files, and our convention is for example `file.c.o` or `file.s.o`. So by cross-referencing `file.s` with our file tree... we could find the full path. Or we could search through the Makefile.

Because that's so weird and potentially unique to our project, it would be fine to just provide a JSON or something of the mappings for symbol name to relative output path like:
```jsonc
{
  "outputPaths": {
    "80ABCDEF": "asm/my_module/my_source.c.s",
    // etc
  }
}
```

This JSON could contain more metadata, if necessary, to reduce or eliminate the need to parse a map file.

**Comment by ribbanya** [comment]
It could also be SQLite and have more information per address, like
Addr|Name|OutputPath
-|-|-
80ABCDEF|myFunction|asm/my_module/my_source.c.s

I don't know what kind of input data is idiomatic of Rust.

**Comment by InusualZ** [comment]
Yeah, I think a JSON file would be the simplest implementation. I will ping you on Discord if I need any more detail.

**Comment by ribbanya** [comment]
Thanks for looking at this!

Here's some sample input: https://gist.github.com/ribbanya/0b000f9e4b00b6954dd950f93687b19a

Let me know if you need any modifications.

**Comment by ribbanya** [comment]
Resolved via e3c9a6ec45158c5f17e96237a0e7aeb2fba9440c. Thanks for implementing this!
